### PR TITLE
More resilient gpg getting

### PR DESCRIPTION
**Pull request by yosifkit**
- move to "high-availability" [subset](https://sks-keyservers.net/overview-of-pools.php#pool_ha).
- do keys one at a time to ensure every key imports

### CODE REVIEW

**Comment by tianon** [comment]
:+1: LGTM

To add a little more context for why this PR is happening, `gpg --recv-keys key1 key2` will only fail (non-zero exit code) if _both_ `key1` and `key2` failed to fetch, so every now and then we get an intermittent keyserver issue that causes one or two keys to fail but the command succeeds so it's baked into our cache with missing keys and the build fails repeatedly after that.  This is the reason for the other part of this where we switch to just high-availability servers and fetch individual keys in a loop. :metal:

**Comment by md5** [comment]
Is there any reason to have `GPG_KEYS` in a separate `ENV` instead of just putting it into the `RUN` statement?

**Comment by tianon** [comment]
One reason is that it makes updating it via "update.sh" simpler (like we do
here).  Another is that it adds an extra layer of simple accountability
making it easy for others to verify the value (ie, scan the list and make
sure it matches what upstream publishes, if/when they do actually publish
the list).

**Comment by tianon** [comment]
Are there good reasons not to do so? (to play a little devil's advocate :smile:)

**Comment by md5** [comment]
Those are good reasons. I didn't scroll all the way down to the `update.sh` changes :see_no_evil: 

I guess I'm just still in a scarcity mindset regarding the number of layers in images.

**Comment by tianon** [comment]
:smile:  Please don't stop making helpful, polite comments like this --
having to be more long-winded about reasons for doing things is ++++ and
usually leads to either more information about the why behind the changes
being publicly posted (which is good) or an improved process in general
(which is even better). :+1:

**Comment by yosifkit** [comment]
```console
Step 6 : RUN set -xe 	&& for key in $GPG_KEYS; do 		gpg --keyserver ha.pool.sks-keyservers.net --recv-keys "$key"; 	done
 ---> Running in b918241a27d6
+ gpg --keyserver ha.pool.sks-keyservers.net --recv-keys 6E4F6AB321FDC07F2C332E3AC2BF0BC433CFC8B3
gpg: directory `/root/.gnupg' created
gpg: new configuration file `/root/.gnupg/gpg.conf' created
gpg: WARNING: options in `/root/.gnupg/gpg.conf' are not yet active during this run
gpg: keyring `/root/.gnupg/secring.gpg' created
gpg: keyring `/root/.gnupg/pubring.gpg' created
gpg: requesting key 33CFC8B3 from hkp server ha.pool.sks-keyservers.net
gpg: /root/.gnupg/trustdb.gpg: trustdb created
gpg: key 33CFC8B3: public key "Ferenc Kovacs <tyrael@php.net>" imported
gpg: Total number processed: 1
gpg:               imported: 1  (RSA: 1)
+ gpg --keyserver ha.pool.sks-keyservers.net --recv-keys 0BD78B5F97500D450838F95DFE857D9A90D90EC1
gpg: requesting key 90D90EC1 from hkp server ha.pool.sks-keyservers.net
gpg: key 90D90EC1: public key "Julien Pauli <jpauli@php.net>" imported
gpg: Total number processed: 1
gpg:               imported: 1  (RSA: 1)
 ---> b1cfedd17311
```

**Comment by tianon** [comment]
Pretty, :heart_eyes:!

File: 5.4/apache/Dockerfile [hunk 1]
```diff
@@ -25,7 +25,11 @@ ENV PHP_EXTRA_BUILD_DEPS apache2-dev
 ENV PHP_EXTRA_CONFIGURE_ARGS --with-apxs2
 ##</autogenerated>##
 
-RUN gpg --keyserver pool.sks-keyservers.net --recv-keys F38252826ACD957EF380D39F2F7956BC5DA04B5D
+ENV GPG_KEYS F38252826ACD957EF380D39F2F7956BC5DA04B5D
+RUN set -xe \
+	&& for key in $GPG_KEYS; do \
+		gpg --keyserver pool.sks-keyservers.net --recv-keys "$key"; \
```
Comment (tianon) [inline 5.4/apache/Dockerfile]: not `ha.`

File: 5.4/fpm/Dockerfile [hunk 2]
```diff
@@ -13,7 +13,11 @@ RUN mkdir -p $PHP_INI_DIR/conf.d
 ENV PHP_EXTRA_CONFIGURE_ARGS --enable-fpm --with-fpm-user=www-data --with-fpm-group=www-data
 ##</autogenerated>##
 
-RUN gpg --keyserver pool.sks-keyservers.net --recv-keys F38252826ACD957EF380D39F2F7956BC5DA04B5D
+ENV GPG_KEYS F38252826ACD957EF380D39F2F7956BC5DA04B5D
+RUN set -xe \
+	&& for key in $GPG_KEYS; do \
+		gpg --keyserver pool.sks-keyservers.net --recv-keys "$key"; \
```
Comment (tianon) [inline 5.4/fpm/Dockerfile]: not `ha.`

File: 5.5/apache/Dockerfile [hunk 3]
```diff
@@ -25,7 +25,11 @@ ENV PHP_EXTRA_BUILD_DEPS apache2-dev
 ENV PHP_EXTRA_CONFIGURE_ARGS --with-apxs2
 ##</autogenerated>##
 
-RUN gpg --keyserver pool.sks-keyservers.net --recv-keys 0BD78B5F97500D450838F95DFE857D9A90D90EC1 0B96609E270F565C13292B24C13C70B87267B52D
+ENV GPG_KEYS 0BD78B5F97500D450838F95DFE857D9A90D90EC1 0B96609E270F565C13292B24C13C70B87267B52D
+RUN set -xe \
+	&& for key in $GPG_KEYS; do \
+		gpg --keyserver pool.sks-keyservers.net --recv-keys "$key"; \
```
Comment (tianon) [inline 5.5/apache/Dockerfile]: not `ha.`

File: 5.5/fpm/Dockerfile [hunk 4]
```diff
@@ -13,7 +13,11 @@ RUN mkdir -p $PHP_INI_DIR/conf.d
 ENV PHP_EXTRA_CONFIGURE_ARGS --enable-fpm --with-fpm-user=www-data --with-fpm-group=www-data
 ##</autogenerated>##
 
-RUN gpg --keyserver pool.sks-keyservers.net --recv-keys 0BD78B5F97500D450838F95DFE857D9A90D90EC1 0B96609E270F565C13292B24C13C70B87267B52D
+ENV GPG_KEYS 0BD78B5F97500D450838F95DFE857D9A90D90EC1 0B96609E270F565C13292B24C13C70B87267B52D
+RUN set -xe \
+	&& for key in $GPG_KEYS; do \
+		gpg --keyserver pool.sks-keyservers.net --recv-keys "$key"; \
```
Comment (tianon) [inline 5.5/fpm/Dockerfile]: not `ha.`

File: 5.6/apache/Dockerfile [hunk 5]
```diff
@@ -25,7 +25,11 @@ ENV PHP_EXTRA_BUILD_DEPS apache2-dev
 ENV PHP_EXTRA_CONFIGURE_ARGS --with-apxs2
 ##</autogenerated>##
 
-RUN gpg --keyserver pool.sks-keyservers.net --recv-keys 6E4F6AB321FDC07F2C332E3AC2BF0BC433CFC8B3 0BD78B5F97500D450838F95DFE857D9A90D90EC1
+ENV GPG_KEYS 6E4F6AB321FDC07F2C332E3AC2BF0BC433CFC8B3 0BD78B5F97500D450838F95DFE857D9A90D90EC1
+RUN set -xe \
+	&& for key in $GPG_KEYS; do \
+		gpg --keyserver pool.sks-keyservers.net --recv-keys "$key"; \
```
Comment (tianon) [inline 5.6/apache/Dockerfile]: not `ha.`

File: 5.6/fpm/Dockerfile [hunk 6]
```diff
@@ -13,7 +13,11 @@ RUN mkdir -p $PHP_INI_DIR/conf.d
 ENV PHP_EXTRA_CONFIGURE_ARGS --enable-fpm --with-fpm-user=www-data --with-fpm-group=www-data
 ##</autogenerated>##
 
-RUN gpg --keyserver pool.sks-keyservers.net --recv-keys 6E4F6AB321FDC07F2C332E3AC2BF0BC433CFC8B3 0BD78B5F97500D450838F95DFE857D9A90D90EC1
+ENV GPG_KEYS 6E4F6AB321FDC07F2C332E3AC2BF0BC433CFC8B3 0BD78B5F97500D450838F95DFE857D9A90D90EC1
+RUN set -xe \
+	&& for key in $GPG_KEYS; do \
+		gpg --keyserver pool.sks-keyservers.net --recv-keys "$key"; \
```
Comment (tianon) [inline 5.6/fpm/Dockerfile]: not `ha.`

**Comment by tianon** [comment]
Those missing `ha.`s are hilariously why the Travis build failed. :+1:

```console
+ gpg --keyserver pool.sks-keyservers.net --recv-keys F38252826ACD957EF380D39F2F7956BC5DA04B5D
gpg: directory `/root/.gnupg' created
gpg: new configuration file `/root/.gnupg/gpg.conf' created
gpg: WARNING: options in `/root/.gnupg/gpg.conf' are not yet active during this run
gpg: keyring `/root/.gnupg/secring.gpg' created
gpg: keyring `/root/.gnupg/pubring.gpg' created
gpg: requesting key 5DA04B5D from hkp server pool.sks-keyservers.net
gpgkeys: key F38252826ACD957EF380D39F2F7956BC5DA04B5D can't be retrieved
gpg: no valid OpenPGP data found.
gpg: Total number processed: 0
INFO[0180] The command [/bin/sh -c set -xe 	&& for key in $GPG_KEYS; do 		gpg --keyserver pool.sks-keyservers.net --recv-keys "$key"; 	done] returned a non-zero code: 2 
```

**Comment by tianon** [comment]
LGTM

**Pull request merged**